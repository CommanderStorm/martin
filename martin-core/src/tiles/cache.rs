use martin_tile_utils::{TileCoord, TileData};
use moka::future::Cache;

/// Tile cache for storing rendered tile data.
#[derive(Clone, Debug)]
pub struct TileCache(Cache<TileCacheKey, TileData>);

impl TileCache {
    /// Creates a new tile cache with the specified maximum size in bytes.
    #[must_use]
    pub fn new(max_size_bytes: u64) -> Self {
        Self(
            Cache::builder()
                .name("tile_cache")
                .weigher(|_key: &TileCacheKey, value: &TileData| -> u32 {
                    value.len().try_into().unwrap_or(u32::MAX)
                })
                .max_capacity(max_size_bytes)
                .build(),
        )
    }

    /// Retrieves a tile from cache if present.
    pub async fn get(
        &self,
        source_id: &str,
        xyz: TileCoord,
        query: Option<&str>,
    ) -> Option<TileData> {
        let key = TileCacheKey::new(source_id, xyz, query);
        let result = self.0.get(&key).await;

        if result.is_some() {
            log::trace!(
                "Tile cache HIT for source={source_id}, xyz={}/{}/{}, query={:?} (entries={}, size={})",
                xyz.z,
                xyz.x,
                xyz.y,
                query,
                self.0.entry_count(),
                self.0.weighted_size()
            );
        } else {
            log::trace!(
                "Tile cache MISS for source={source_id}, xyz={}/{}/{}, query={:?}",
                xyz.z,
                xyz.x,
                xyz.y,
                query
            );
        }

        result
    }

    /// Inserts a tile into the cache.
    pub async fn insert(
        &self,
        source_id: &str,
        xyz: TileCoord,
        query: Option<&str>,
        data: TileData,
    ) {
        let key = TileCacheKey::new(source_id, xyz, query);
        self.0.insert(key, data).await;
    }

    /// Gets a tile from cache or computes it using the provided function.
    pub async fn get_or_insert<F, Fut, E>(
        &self,
        source_id: &str,
        xyz: TileCoord,
        query: Option<&str>,
        compute: F,
    ) -> Result<TileData, E>
    where
        F: FnOnce() -> Fut,
        Fut: Future<Output = Result<TileData, E>>,
    {
        if let Some(data) = self.get(source_id, xyz, query).await {
            return Ok(data);
        }

        let data = compute().await?;
        self.insert(source_id, xyz, query, data.clone()).await;
        Ok(data)
    }

    /// Invalidates all cached tiles for a specific source.
    pub async fn invalidate_source(&self, source_id: &str) {
        let source_id_owned = source_id.to_string();
        self.0
            .invalidate_entries_if(move |key, _| key.source_id == source_id_owned)
            .expect("invalidate_entries_if predicate should not error");
        log::info!("Invalidated tile cache for source: {source_id}");
    }

    /// Invalidates all cached tiles.
    pub async fn invalidate_all(&self) {
        self.0.invalidate_all();
        log::info!("Invalidated all tile cache entries");
    }

    /// Returns the number of cached entries.
    #[must_use]
    pub fn entry_count(&self) -> u64 {
        self.0.entry_count()
    }

    /// Returns the total size of cached data in bytes.
    #[must_use]
    pub fn weighted_size(&self) -> u64 {
        self.0.weighted_size()
    }
}

/// Optional wrapper for `TileCache`.
pub type OptTileCache = Option<TileCache>;

/// Constant representing no tile cache configuration.
pub const NO_TILE_CACHE: OptTileCache = None;

/// Cache key for tile data.
#[derive(Debug, Hash, PartialEq, Eq, Clone)]
struct TileCacheKey {
    source_id: String,
    xyz: TileCoord,
    query: Option<String>,
}

impl TileCacheKey {
    fn new(source_id: &str, xyz: TileCoord, query: Option<&str>) -> Self {
        Self {
            source_id: source_id.to_string(),
            xyz,
            query: query.map(ToString::to_string),
        }
    }
}
