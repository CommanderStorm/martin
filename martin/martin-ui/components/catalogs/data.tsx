"use client";

import {
	Database,
	Eye,
	Globe,
	ImageIcon,
	Layers,
	Palette,
	Search,
	Type,
} from "lucide-react";
import { ErrorState, InlineErrorState } from "@/components/error/error-state";
import { CatalogSkeleton } from "@/components/loading/catalog-skeleton";
import { LoadingSpinner } from "@/components/loading/loading-spinner";
import { Badge } from "@/components/ui/badge";
import { Button } from "@/components/ui/button";
import {
	Card,
	CardContent,
	CardDescription,
	CardHeader,
	CardTitle,
} from "@/components/ui/card";
import { Input } from "@/components/ui/input";
import { Tooltip, TooltipContent, TooltipTrigger } from "../ui/tooltip";

interface DataSource {
	id: string;
	name: string;
	type: string;
	description: string;
	layers: number;
	lastUpdated: string;
	size: string;
}

interface DataCatalogProps {
	dataSources: DataSource[];
	searchQuery: string;
	onSearchChangeAction: (query: string) => void;
	isLoading?: boolean;
	isSearching?: boolean;
	error?: Error | null;
	searchError?: Error | null;
	onRetry?: () => void;
	onRetrySearch?: () => void;
	isRetrying?: boolean;
}

export function DataCatalog({
	dataSources,
	searchQuery,
	onSearchChangeAction,
	isLoading = false,
	isSearching = false,
	error = null,
	searchError = null,
	onRetry,
	onRetrySearch,
	isRetrying = false,
}: DataCatalogProps) {
	if (isLoading) {
		return (
			<CatalogSkeleton
				title="Data Sources Catalog"
				description="Manage and explore all available tile sources, sprites, and fonts"
			/>
		);
	}

	if (error) {
		return (
			<ErrorState
				title="Failed to Load Data Catalog"
				description="Unable to fetch data sources from the server"
				error={error}
				onRetry={onRetry}
				isRetrying={isRetrying}
				variant="server"
				showDetails={true}
			/>
		);
	}

	const filteredDataSources = dataSources.filter(
		(source) =>
			source.name.toLowerCase().includes(searchQuery.toLowerCase()) ||
			source.type.toLowerCase().includes(searchQuery.toLowerCase()),
	);

	const getIcon = (type: string) => {
		switch (type) {
			case "Vector Tiles":
				return <Layers className="w-5 h-5 text-primary" />;
			case "Raster Tiles":
				return <ImageIcon className="w-5 h-5 text-primary" />;
			case "Sprites":
				return <Globe className="w-5 h-5 text-primary" />;
			case "Fonts":
				return <Type className="w-5 h-5 text-primary" />;
			default:
				return <Database className="w-5 h-5 text-primary" />;
		}
	};

	return (
		<div className="space-y-6">
			<div className="flex items-center justify-between">
				<div>
					<h2 className="text-2xl font-bold text-foreground">
						Data Sources Catalog
					</h2>
					<p className="text-muted-foreground">
						Manage and explore all available tile sources, sprites, and fonts
					</p>
				</div>
				<div className="relative">
					<Search className="absolute left-3 top-1/2 transform -translate-y-1/2 text-gray-400 w-4 h-4" />
					<Input
						placeholder="Search data sources..."
						value={searchQuery}
						onChange={(e) => onSearchChangeAction(e.target.value)}
						className="pl-10 w-64 bg-card"
					/>
					{isSearching && (
						<div className="absolute right-3 top-1/2 transform -translate-y-1/2">
							<LoadingSpinner size="sm" />
						</div>
					)}
				</div>
			</div>

			{searchError && (
				<InlineErrorState
					message="Search failed. Please try again."
					onRetry={onRetrySearch}
					variant="network"
				/>
			)}

			{isSearching ? (
				<div className="flex items-center justify-center py-12">
					<div className="text-center">
						<LoadingSpinner size="lg" className="mx-auto mb-4" />
						<p className="text-muted-foreground">Searching data sources...</p>
					</div>
				</div>
			) : (
				<div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-6">
					{filteredDataSources.map((source) => (
						<Card key={source.id} className="hover:shadow-lg transition-shadow">
							<CardHeader>
								<div className="flex items-center justify-between">
									<div className="flex items-center space-x-2">
										{getIcon(source.type)}
										<CardTitle className="text-lg">{source.name}</CardTitle>
									</div>
									<Badge variant="secondary">{source.type}</Badge>
								</div>
								<CardDescription>{source.description}</CardDescription>
							</CardHeader>
							<CardContent>
								<div className="space-y-2 text-sm text-muted-foreground">
									<div className="flex justify-between">
										<span>Layers:</span>
										<span>{source.layers}</span>
									</div>
									<div className="flex justify-between">
										<span>Size:</span>
										<span>{source.size}</span>
									</div>
									<div className="flex justify-between">
										<span>Updated:</span>
										<span>{source.lastUpdated}</span>
									</div>
								</div>
								<div className="flex space-x-2 mt-4">
  								<Tooltip>
                    <TooltipTrigger className="flex flex-1" asChild>
     									<Button
      										size="sm"
      										variant="outline"
      										className="flex-1 bg-transparent"
      										disabled
     									>
      										<Eye className="w-4 h-4 mr-2" />
      										Inspect
     									</Button>
                    </TooltipTrigger>
  									<TooltipContent>
                      <p>Not currently implemented in the frontend</p>
                    </TooltipContent>
                  </Tooltip>
  								<Tooltip>
                    <TooltipTrigger className="flex flex-1" asChild>
   									<Button
    										size="sm"
    										className="flex-1 bg-primary hover:bg-purple-700"
    										disabled
    								>
    										<Palette className="w-4 h-4 mr-2" />
    										Style
   									</Button>
    								</TooltipTrigger>
  									<TooltipContent>
                      <p>Not currently implemented in the frontend</p>
                    </TooltipContent>
                  </Tooltip>
								</div>
							</CardContent>
						</Card>
					))}
				</div>
			)}

			{!isSearching &&
				filteredDataSources.length === 0 &&
				searchQuery &&
				!searchError && (
					<div className="text-center py-12">
						<p className="text-muted-foreground">
							No data sources found matching "{searchQuery}"
						</p>
					</div>
				)}
		</div>
	);
}
