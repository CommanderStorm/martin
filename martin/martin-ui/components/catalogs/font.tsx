import { Eye, Search, Type } from "lucide-react";
import Link from "next/link";
import { ErrorState } from "@/components/error/error-state";
import { CatalogSkeleton } from "@/components/loading/catalog-skeleton";
import { Badge } from "@/components/ui/badge";
import { Button } from "@/components/ui/button";
import { Card, CardContent, CardDescription, CardHeader, CardTitle } from "@/components/ui/card";
import { Input } from "@/components/ui/input";
import type { Font } from "@/lib/types";
import { DisabledNonInteractiveButton } from "../ui/disabledNonInteractiveButton";
import { Tooltip, TooltipContent, TooltipTrigger } from "../ui/tooltip";
import { CopyLinkButton } from "../ui/copy-link-button";
import { useToast } from "../ui/use-toast";

interface FontCatalogProps {
  fonts?: { [name: string]: Font };
  searchQuery?: string;
  onSearchChangeAction?: (query: string) => void;
  isLoading?: boolean;
  error?: Error | null;
  onRetry?: () => void;
  isRetrying?: boolean;
}

export function FontCatalog({
  fonts,
  searchQuery = "",
  onSearchChangeAction = () => {},
  isLoading,
  error = null,
  onRetry,
  isRetrying = false,
}: FontCatalogProps) {
  const { toast } = useToast();
  if (isLoading) {
    return <CatalogSkeleton description="Preview all available font assets" title="Font Catalog" />;
  }

  if (error) {
    return (
      <ErrorState
        description="Unable to fetch font catalog from the server"
        error={error}
        isRetrying={isRetrying}
        onRetry={onRetry}
        showDetails={true}
        title="Failed to Load Fonts"
        variant="server"
      />
    );
  }

  const filteredFonts = Object.entries(fonts || {}).filter(
    ([name, font]) =>
      name.toLowerCase().includes(searchQuery.toLowerCase()) ||
      font.family.toLowerCase().includes(searchQuery.toLowerCase()) ||
      font.style.toLowerCase().includes(searchQuery.toLowerCase()),
  );

  return (
    <div className="space-y-6">
      <div className="flex items-center justify-between">
        <div>
          <h2 className="text-2xl font-bold text-foreground">Font Catalog</h2>
          <p className="text-muted-foreground">Preview all available font assets</p>
        </div>
        <div className="relative">
          <Search className="absolute left-3 top-1/2 transform -translate-y-1/2 text-gray-400 w-4 h-4" />
          <Input
            className="pl-10 w-64 bg-card"
            onChange={(e) => onSearchChangeAction(e.target.value)}
            placeholder="Search fonts..."
            value={searchQuery}
          />
        </div>
      </div>

      <div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-6">
        {filteredFonts.map(([name, font]) => (
          <Card className="hover:shadow-lg transition-shadow" key={name}>
            <CardHeader>
              <div className="flex items-center justify-between">
                <div className="flex items-center space-x-2">
                  <Type className="w-5 h-5 text-primary" />
                  <CardTitle className="text-lg">{name}</CardTitle>
                </div>
                {font.format && (
                  <Badge className="uppercase" variant="secondary">
                    {font.format}
                  </Badge>
                )}
              </div>
              <CardDescription>
                Family: {font.family} â€¢ Style: {font.style}
              </CardDescription>
            </CardHeader>
            <CardContent>
              <div className="space-y-4">
                <div className="p-3 bg-gray-50 text-gray-900 rounded-lg">
                  <p className="text-sm font-medium mb-2 text-gray-900">Preview:</p>
                  <Tooltip>
                    <TooltipTrigger className="cursor-help">
                      <p
                        className="text-base text-gray-900 blur-sm animate-pulse"
                        style={{ fontFamily: font.family, fontWeight: 500 }}
                      >
                        The quick brown fox jumps over the lazy dog
                      </p>
                    </TooltipTrigger>
                    <TooltipContent>
                      <p>Not currently implemented in the frontend</p>
                    </TooltipContent>
                  </Tooltip>
                </div>
                <div className="space-y-2 text-sm text-muted-foreground">
                  <div className="flex justify-between">
                    <span>Glyph count:</span>
                    <span>{font.glyphs}</span>
                  </div>
                </div>
                <div className="flex space-x-2">
                  <CopyLinkButton
                    className="flex-1 bg-transparent"
                    size="sm"
                    variant="outline"
                    link={`/font/${name}/{range}`}
                    toastMessage="Font link copied!"
                  />
                  <Tooltip>
                    <TooltipTrigger className="flex-1 flex cursor-help">
                      <DisabledNonInteractiveButton className="flex-1 grow" disabled size="sm">
                        <Eye className="w-4 h-4 mr-2" />
                        Details
                      </DisabledNonInteractiveButton>
                    </TooltipTrigger>
                    <TooltipContent>
                      <p>Not currently implemented in the frontend</p>
                    </TooltipContent>
                  </Tooltip>
                </div>
              </div>
            </CardContent>
          </Card>
        ))}
      </div>

      {filteredFonts.length === 0 && (
        <div className="text-center py-12">
          <p className="text-muted-foreground mb-2">
            {searchQuery
              ? <>No fonts found matching "{searchQuery}".</>
              : <>No fonts found.</>
            }
          </p>
          <Button
            asChild
            variant="link"
            size="lg"
          >
            <Link
              href="https://maplibre.org/martin/sources-fonts.html"
              target="_blank"
              rel="noopener noreferrer"
            >
              Learn how to configure Fonts
            </Link>
          </Button>
        </div>
      )}
    </div>
  );
}
