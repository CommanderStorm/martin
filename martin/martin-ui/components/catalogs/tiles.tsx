"use client";

import {
	Database,
	Eye,
	ImageIcon,
	Layers,
	Palette,
	Search,
} from "lucide-react";
import { ErrorState } from "@/components/error/error-state";
import { CatalogSkeleton } from "@/components/loading/catalog-skeleton";
import { Badge } from "@/components/ui/badge";
import {
	Card,
	CardContent,
	CardDescription,
	CardHeader,
	CardTitle,
} from "@/components/ui/card";
import { Input } from "@/components/ui/input";
import type { TileSource } from "@/lib/types";
import { DisabledNonInteractiveButton } from "../ui/disabledNonInteractiveButton";
import { Tooltip, TooltipContent, TooltipTrigger } from "../ui/tooltip";

interface TilesCatalogProps {
	tileSources: { [tile_id: string]: TileSource } | null;
	searchQuery: string;
	onSearchChangeAction: (query: string) => void;
	isLoading?: boolean;
	error?: Error | null;
	onRetry?: () => void;
	isRetrying?: boolean;
}

export function TilesCatalog({
	tileSources,
	searchQuery,
	onSearchChangeAction,
	isLoading,
	error = null,
	onRetry,
	isRetrying = false,
}: TilesCatalogProps) {
	if (isLoading) {
		return (
			<CatalogSkeleton
				title="Tiles Sources Catalog"
				description="Explore all available tile sources, sprites, and fonts"
			/>
		);
	}

	if (error) {
		return (
			<ErrorState
				title="Failed to Load Tiles Catalog"
				description="Unable to fetch tiles sources from the server"
				error={error}
				onRetry={onRetry}
				isRetrying={isRetrying}
				variant="server"
				showDetails={true}
			/>
		);
	}

	const filteredTileSources = Object.entries(tileSources || {}).filter(
		([name, source]) =>
			name.toLowerCase().includes(searchQuery.toLowerCase()) ||
			source.name?.toLowerCase().includes(searchQuery.toLowerCase()) ||
			source.description?.toLowerCase().includes(searchQuery.toLowerCase()) ||
			source.attribution?.toLowerCase().includes(searchQuery.toLowerCase()) ||
			source.content_encoding?.toLowerCase().includes(searchQuery.toLowerCase()) ||
			source.content_type?.toLowerCase().includes(searchQuery.toLowerCase()),
	);

	const getIcon = (content_type: string) => {
	if (content_type.startsWith("image/"))
	return <ImageIcon className="w-5 h-5 text-primary" />;
	if (content_type === "application/x-protobuf")
	return <Layers className="w-5 h-5 text-primary" />;
				return <Database className="w-5 h-5 text-primary" />;
	};

	return (
		<div className="space-y-6">
			<div className="flex items-center justify-between">
				<div>
					<h2 className="text-2xl font-bold text-foreground">
						Tiles Sources Catalog
					</h2>
					<p className="text-muted-foreground">
						Explore all available tile sources, sprites, and fonts
					</p>
				</div>
				<div className="relative">
					<Search className="absolute left-3 top-1/2 transform -translate-y-1/2 text-gray-400 w-4 h-4" />
					<Input
						placeholder="Search tiles sources..."
						value={searchQuery}
						onChange={(e) => onSearchChangeAction(e.target.value)}
						className="pl-10 w-64 bg-card"
					/>
				</div>
			</div>

			<div>
				<div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-6">
					{filteredTileSources.map(([name, source]) => (
						<Card key={name} className="hover:shadow-lg transition-shadow">
							<CardHeader>
								<div className="flex items-center justify-between">
									<div className="flex items-center space-x-2">
										{getIcon(source.content_type)}
										<CardTitle className="text-lg">{source.name}</CardTitle>
									</div>
									<Badge variant="secondary">{source.content_type}</Badge>
								</div>
								<CardDescription>{source.description}</CardDescription>
							</CardHeader>
							<CardContent>
								<div className="space-y-2 text-sm text-muted-foreground">
									{source.layers && (
										<div className="flex justify-between">
											<span>Layers:</span>
											<span>{source.layers}</span>
										</div>
									)}
								</div>
								<div className="flex space-x-2 mt-4">
									<Tooltip>
										<TooltipTrigger className="flex flex-1">
											<DisabledNonInteractiveButton
												size="sm"
												variant="outline"
												className="flex-1 bg-transparent"
											>
												<Eye className="w-4 h-4 mr-2" />
												Inspect
											</DisabledNonInteractiveButton>
										</TooltipTrigger>
										<TooltipContent>
											<p>Not currently implemented in the frontend</p>
										</TooltipContent>
									</Tooltip>
									<Tooltip>
										<TooltipTrigger className="flex flex-1">
											<DisabledNonInteractiveButton
												size="sm"
												className="flex-1"
											>
												<Palette className="w-4 h-4 mr-2" />
												Style
											</DisabledNonInteractiveButton>
										</TooltipTrigger>
										<TooltipContent>
											<p>Not currently implemented in the frontend</p>
										</TooltipContent>
									</Tooltip>
								</div>
							</CardContent>
						</Card>
					))}
				</div>
			</div>

			{filteredTileSources.length === 0 && searchQuery && (
				<div className="text-center py-12">
					<p className="text-muted-foreground">
						No tile sources found matching "{searchQuery}"
					</p>
				</div>
			)}
		</div>
	);
}
