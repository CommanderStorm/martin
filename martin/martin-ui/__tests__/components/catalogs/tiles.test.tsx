import { within } from "@testing-library/dom";
import { fireEvent, render, screen } from "@testing-library/react";
import type React from "react";
import { TilesCatalog } from "@/components/catalogs/tiles";
import type { TileSource } from "@/lib/types";

// Mock all dependencies
jest.mock("@/components/error/error-state", () => ({
	ErrorState: ({
		title,
		description,
	}: {
		title: string;
		description: string;
	}) => (
		<div data-testid="error-state">
			<div data-testid="error-title">{title}</div>
			<div data-testid="error-description">{description}</div>
		</div>
	),
}));

jest.mock("@/components/loading/catalog-skeleton", () => ({
	CatalogSkeleton: ({
		title,
		description,
	}: {
		title: string;
		description: string;
	}) => (
		<div data-testid="catalog-skeleton">
			<div data-testid="skeleton-title">{title}</div>
			<div data-testid="skeleton-description">{description}</div>
		</div>
	),
}));

// Mock UI components to avoid tooltip provider issues
jest.mock("@/components/ui/tooltip", () => ({
	Tooltip: ({ children }: { children: React.ReactNode }) => (
		<div>{children}</div>
	),
	TooltipTrigger: ({ children }: { children: React.ReactNode }) => (
		<div>{children}</div>
	),
	TooltipContent: ({ children }: { children: React.ReactNode }) => (
		<div data-testid="tooltip-content">{children}</div>
	),
}));

jest.mock("@/components/ui/button", () => ({
	Button: ({ children, ...props }: any) => (
		<button {...props}>{children}</button>
	),
}));

jest.mock("@/components/ui/badge", () => ({
	Badge: ({ children, ...props }: any) => (
		<span data-testid="badge" {...props}>
			{children}
		</span>
	),
}));

jest.mock("@/components/ui/input", () => ({
	Input: (props: any) => <input {...props} />,
}));

jest.mock("@/components/ui/card", () => ({
	Card: ({ children, ...props }: any) => <div {...props}>{children}</div>,
	CardContent: ({ children, ...props }: any) => (
		<div data-testid="card-content" {...props}>
			{children}
		</div>
	),
	CardHeader: ({ children, ...props }: any) => (
		<div data-testid="card-header" {...props}>
			{children}
		</div>
	),
	CardTitle: ({ children, ...props }: any) => (
		<div data-testid="card-title" {...props}>
			{children}
		</div>
	),
	CardDescription: ({ children, ...props }: any) => (
		<div data-testid="card-description" {...props}>
			{children}
		</div>
	),
}));

jest.mock("@/components/ui/disabledNonInteractiveButton", () => ({
	DisabledNonInteractiveButton: ({ children, ...props }: any) => (
		<button {...props} disabled>
			{children}
		</button>
	),
}));

jest.mock("lucide-react", () => ({
	Database: () => <div data-testid="database-icon">DB</div>,
	Eye: () => <div data-testid="eye-icon">Eye</div>,
	ImageIcon: () => <div data-testid="image-icon">Image</div>,
	Layers: () => <div data-testid="layers-icon">Layers</div>,
	Palette: () => <div data-testid="palette-icon">Palette</div>,
	Search: () => <div data-testid="search-icon">Search</div>,
}));

describe("TilesCatalog Component", () => {
	const mockTileSources: { [tile_id: string]: TileSource } = {
		source1: {
			name: "Test Source 1",
			description: "Description for test source 1",
			content_type: "application/x-protobuf",
			attribution: "Test attribution 1",
			layerCount: 5,
		},
		source2: {
			name: "Test Source 2",
			description: "Description for test source 2",
			content_type: "image/png",
			attribution: "Test attribution 2",
			layerCount: 3,
		},
		source3: {
			name: "Test Source 3",
			description: "Description for test source 3",
			content_type: "application/json",
			attribution: "Test attribution 3",
			layerCount: 7,
		},
	};

	const defaultProps = {
		tileSources: mockTileSources,
		searchQuery: "",
		onSearchChangeAction: jest.fn(),
		isLoading: false,
		error: null,
		onRetry: jest.fn(),
		isRetrying: false,
	};

	it("renders loading skeleton when isLoading is true", () => {
		render(<TilesCatalog {...defaultProps} isLoading={true} />);
		expect(screen.getByTestId("catalog-skeleton")).toBeInTheDocument();
	});

	it("renders error state when there is an error", () => {
		const error = new Error("Test error");
		render(<TilesCatalog {...defaultProps} error={error} />);
		expect(screen.getByTestId("error-state")).toBeInTheDocument();
		expect(screen.getByTestId("error-title").textContent).toBe(
			"Failed to Load Tiles Catalog",
		);
	});

	it("renders tile sources correctly", () => {
		render(<TilesCatalog {...defaultProps} />);
		expect(screen.getByText("Tiles Sources Catalog")).toBeInTheDocument();

		// Get all card descriptions - these contain the source names
		const descriptions = screen.getAllByTestId("card-description");

		// Check that each of our expected source names exists somewhere
		expect(
			descriptions.some((el) => el.textContent?.includes("Test Source 1")),
		).toBe(true);
		expect(
			descriptions.some((el) => el.textContent?.includes("Test Source 2")),
		).toBe(true);
		expect(
			descriptions.some((el) => el.textContent?.includes("Test Source 3")),
		).toBe(true);
	});

	it("filters tile sources based on search query", () => {
		render(<TilesCatalog {...defaultProps} searchQuery="source 2" />);

		// Get all card descriptions - these contain the source names
		const descriptions = screen.getAllByTestId("card-description");

		// Should have only one source that matches
		expect(descriptions.length).toBe(1);
		expect(descriptions[0].textContent).toContain("Test Source 2");

		// Grid should only contain the source2 component
		const gridItems = screen.getAllByTestId("card-header");
		expect(gridItems.length).toBe(1);
	});

	it("shows no results message when search has no matches", () => {
		render(<TilesCatalog {...defaultProps} searchQuery="nonexistent" />);
		expect(
			screen.getByText(/No tile sources found matching "nonexistent"/i),
		).toBeInTheDocument();

		// Should not render any cards
		const gridItems = screen.queryAllByTestId("card-header");
		expect(gridItems.length).toBe(0);
	});

	it("calls onSearchChangeAction when search input changes", () => {
		render(<TilesCatalog {...defaultProps} />);
		const searchInput = screen.getByPlaceholderText("Search tiles sources...");

		fireEvent.change(searchInput, { target: { value: "new search" } });
		expect(defaultProps.onSearchChangeAction).toHaveBeenCalledWith(
			"new search",
		);
	});

	it("renders different icons based on content_type", () => {
		render(<TilesCatalog {...defaultProps} />);

		// Each icon type should be present exactly once (one for each source)
		const layersIcon = screen.getAllByTestId("layers-icon");
		const imageIcon = screen.getAllByTestId("image-icon");
		const databaseIcon = screen.getAllByTestId("database-icon");

		expect(layersIcon.length).toBe(1);
		expect(imageIcon.length).toBe(1);
		expect(databaseIcon.length).toBe(1);
	});
});
