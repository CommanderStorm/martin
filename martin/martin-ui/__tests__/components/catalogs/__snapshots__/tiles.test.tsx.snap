// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`TilesCatalog Component matches snapshot for loaded state with mock data 1`] = `
<DocumentFragment>
  <div
    class="space-y-6"
  >
    <div
      class="flex items-center justify-between"
    >
      <div>
        <h2
          class="text-2xl font-bold text-foreground"
        >
          Tiles Sources Catalog
        </h2>
        <p
          class="text-muted-foreground"
        >
          Explore all available tile sources
        </p>
      </div>
      <div
        class="relative"
      >
        <div
          data-testid="search-icon"
        >
          Search
        </div>
        <input
          class="pl-10 w-64 bg-card"
          placeholder="Search tiles sources..."
          value=""
        />
      </div>
    </div>
    <div>
      <div
        class="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-6"
      >
        <div
          class="hover:shadow-lg transition-shadow"
        >
          <div
            data-testid="card-header"
          >
            <div
              class="flex items-center justify-between"
            >
              <div
                class="flex items-center space-x-2"
              >
                <div
                  data-testid="layers-icon"
                >
                  Layers
                </div>
                <div
                  class="text-lg"
                  data-testid="card-title"
                >
                  source1
                </div>
              </div>
              <span
                data-testid="badge"
                variant="secondary"
              >
                application/x-protobuf
              </span>
            </div>
            <div
              data-testid="card-description"
            >
              Test Source 1
              <br />
              Description for test source 1
            </div>
          </div>
          <div
            data-testid="card-content"
          >
            <div
              class="space-y-2 text-sm text-muted-foreground"
            >
              <div
                class="flex justify-between"
              >
                <span>
                  Layers:
                </span>
                <span>
                  5
                </span>
              </div>
            </div>
            <div
              class="flex space-x-2 mt-4"
            >
              <div>
                <div>
                  <button
                    class="flex-1 bg-transparent"
                    disabled=""
                    variant="outline"
                  >
                    <div
                      data-testid="eye-icon"
                    >
                      Eye
                    </div>
                    Inspect
                  </button>
                </div>
                <div
                  data-testid="tooltip-content"
                >
                  <p>
                    Not currently implemented in the frontend
                  </p>
                </div>
              </div>
              <div>
                <div>
                  <button
                    class="flex-1"
                    disabled=""
                  >
                    <div
                      data-testid="palette-icon"
                    >
                      Palette
                    </div>
                    Style
                  </button>
                </div>
                <div
                  data-testid="tooltip-content"
                >
                  <p>
                    Not currently implemented in the frontend
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="hover:shadow-lg transition-shadow"
        >
          <div
            data-testid="card-header"
          >
            <div
              class="flex items-center justify-between"
            >
              <div
                class="flex items-center space-x-2"
              >
                <div
                  data-testid="image-icon"
                >
                  Image
                </div>
                <div
                  class="text-lg"
                  data-testid="card-title"
                >
                  source2
                </div>
              </div>
              <span
                data-testid="badge"
                variant="secondary"
              >
                image/png
              </span>
            </div>
            <div
              data-testid="card-description"
            >
              Test Source 2
              <br />
              Description for test source 2
            </div>
          </div>
          <div
            data-testid="card-content"
          >
            <div
              class="space-y-2 text-sm text-muted-foreground"
            >
              <div
                class="flex justify-between"
              >
                <span>
                  Layers:
                </span>
                <span>
                  3
                </span>
              </div>
            </div>
            <div
              class="flex space-x-2 mt-4"
            >
              <div>
                <div>
                  <button
                    class="flex-1 bg-transparent"
                    disabled=""
                    variant="outline"
                  >
                    <div
                      data-testid="eye-icon"
                    >
                      Eye
                    </div>
                    Inspect
                  </button>
                </div>
                <div
                  data-testid="tooltip-content"
                >
                  <p>
                    Not currently implemented in the frontend
                  </p>
                </div>
              </div>
              <div>
                <div>
                  <button
                    class="flex-1"
                    disabled=""
                  >
                    <div
                      data-testid="palette-icon"
                    >
                      Palette
                    </div>
                    Style
                  </button>
                </div>
                <div
                  data-testid="tooltip-content"
                >
                  <p>
                    Not currently implemented in the frontend
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="hover:shadow-lg transition-shadow"
        >
          <div
            data-testid="card-header"
          >
            <div
              class="flex items-center justify-between"
            >
              <div
                class="flex items-center space-x-2"
              >
                <div
                  data-testid="database-icon"
                >
                  DB
                </div>
                <div
                  class="text-lg"
                  data-testid="card-title"
                >
                  source3
                </div>
              </div>
              <span
                data-testid="badge"
                variant="secondary"
              >
                application/json
              </span>
            </div>
            <div
              data-testid="card-description"
            >
              Test Source 3
              <br />
              Description for test source 3
            </div>
          </div>
          <div
            data-testid="card-content"
          >
            <div
              class="space-y-2 text-sm text-muted-foreground"
            >
              <div
                class="flex justify-between"
              >
                <span>
                  Layers:
                </span>
                <span>
                  7
                </span>
              </div>
            </div>
            <div
              class="flex space-x-2 mt-4"
            >
              <div>
                <div>
                  <button
                    class="flex-1 bg-transparent"
                    disabled=""
                    variant="outline"
                  >
                    <div
                      data-testid="eye-icon"
                    >
                      Eye
                    </div>
                    Inspect
                  </button>
                </div>
                <div
                  data-testid="tooltip-content"
                >
                  <p>
                    Not currently implemented in the frontend
                  </p>
                </div>
              </div>
              <div>
                <div>
                  <button
                    class="flex-1"
                    disabled=""
                  >
                    <div
                      data-testid="palette-icon"
                    >
                      Palette
                    </div>
                    Style
                  </button>
                </div>
                <div
                  data-testid="tooltip-content"
                >
                  <p>
                    Not currently implemented in the frontend
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TilesCatalog Component matches snapshot for loading state 1`] = `
<DocumentFragment>
  <div
    data-testid="catalog-skeleton"
  >
    <div
      data-testid="skeleton-title"
    >
      Tiles Sources Catalog
    </div>
    <div
      data-testid="skeleton-description"
    >
      Explore all available tile sources
    </div>
  </div>
</DocumentFragment>
`;
